src/wbase/Task.h

```diff
@@ -295,6 +295,12 @@ class Task : public util::CommandForThreadPool {
     /// system in order to see what MySQL queries are being executed by tasks.
     void setMySqlThreadId(unsigned long id) { _mysqlThreadId.store(id); }
 
+    /// Return true if this task was already booted.
+    bool setBooted() { return _booted.exchange(true); }
+
+    /// Return true if the task was booted.
+    bool isBooted() { return _booted; }
+
```

Comment:
I'm not sure that simply adding yet another data member would be the right approach. I think that, the problem may need to be considered in a more general context of the class design. In particular, this class `wbase::Task` is a state machine that goes through a series of state transitions, depending on what's happening during query processing. The class has the following data members representing its state:
```
class Task {
private:
    std::atomic<TaskState> _state{TaskState::CREATED};
};
```
Where:
```
/**
 * The class State represents transient states of the worker-side tasks.
 * @note This class and the relevant functions are put into this header to
 * allow the compile-time (only) dependency onto this type from other modules
 * without needing to link against the current module's library.
 * Also note a choice of the underlying type which is meant to allow sending
 * values of the type as numeric attributes in the Protobuf messages w/o
 * introducing an additional (Protobuf) representation for those, or converting
 * the values to strings and vs.
 */
enum class TaskState : std::uint64_t {
    CREATED = 0,
    QUEUED,
    STARTED,
    EXECUTING_QUERY,
    READING_DATA,
    FINISHED
};
```
My impression is that "booting" is a significant event in the lifespan of the task, and as such it would need to be incorporated as a state. On the other hand, "booting" is not quite the state transition as it may mean one of 3 things:
1. merely detaching the task from a scheduler into the "free run" world
2. moving the task to some other low-priority scheduler ("snail", etc.)
3. or, canceling the task and (possibly queueing it into some low-priority) queue.

My understanding is that you're attempting to implement case number 2. Is this correct?

I also don't think you need to introduce these extra methods `setBooted()` and `isBooted()` since these will create redundancy and non-orthogonality. We already have the following methods which do effectively exactly what you wanted to achieve with the "boot" methods:
```
class Task {
public:
    void setTaskScheduler(TaskScheduler::Ptr const& scheduler) { _taskScheduler = scheduler; }
    TaskScheduler::Ptr getTaskScheduler() { return _taskScheduler.lock(); }
};
```
If you're implementing either scenario 1 (detaching from any scheduler) or 3 (canceling the task and re-queueing it) then you're going to reset the schedule to `nullptr`. If you're implementing scenario 2 (moving to "snail") then you are going to replace a value of the pointer with something else.

What I have not found in the new code is a timestamp when either of these significant events (1, 2, or 3) of making changes to the scheduler might happen. This could be done by modifying the implementation of the method `Task::setTaskScheduler()` to be more complex:
```
class Task {
public:
    void setTaskScheduler(TaskScheduler::Ptr const& scheduler) {
        _taskScheduler = scheduler;
       _schedChangeTime  =   std::chrono::system_clock::now();
    }

private:
    /// A timestamp when the last change of the scheduler was made
    std::chrono::system_clock::time_point _schedChangeTime;
};
```
This timestamp along with a value of the scheduler pointer is _exactly_ what is needed for the monitoring system to what's going on with the long-run tasks.

Reply:
The problem is that "_booted" overlaps with other states, at least with EXECUTING_QUERY and possibly READING_DATA. Technically, what's happening is closest to your option 1, but the booted tasks should really be assigned to the snailScan so that snailScan is aware of their existance, and can then use them to make decisions about starting new Tasks from its queue. I don't think I'm going to go that far in this ticket, but it's probably what should happen.

I think keeping track of a task being booted is also important rguardless of the state of the task. What I mean is that as long as the system is holding onto the `Task`, we care about it being booted, even when it is `FINISHED`. When trying to figure out why a Task had issues, having been booted is something we probably want to know. Your proposed solution falls a little short in that every time a task get scheduled, it wipes out the previous timestamp, it also means there needs to be a separate timestamp for each TaskState change.
I think a better stratedgy would be to make a list of optional events for the task with timestamps. This would allow multiple scheduler changes to be recorded, as well multiple boot events. In most cases, this list would be empty. If we do decide to go with @fritzm 's idea of waterfall booting of Tasks, this would work reasonably well.